### PR TITLE
Feature/lock time 30

### PR DESCRIPTION
**Pull request by danoctavian**

### CODE REVIEW

**Review by roxdanila — CHANGES_REQUESTED** [review]

**Comment by roxdanila** [comment]
@danoctavian check out the failing test, looks related

File: contracts/modules/staking/PooledStaking.sol [hunk 1]
```diff
@@ -405,6 +405,13 @@ contract PooledStaking is MasterAware, IPooledStaking {
     uint _insertAfter // unstake request id after which the new unstake request will be inserted
   ) external whenNotPausedAndInitialized onlyMember {
 
+    bytes32 stagePosition = LOCK_TIME_MIGRATION_STAGE_POSITION;
+    uint migrationStage = 0;
+    assembly {
+      migrationStage := sload(stagePosition)
+    }
+    require(migrationStage == 2, "PooledStaking: Migration in progress");
+
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply contracts/modules/staking/PooledStaking.sol:414]: with check:
{ gas: 188371 }
without check:
{ gas: 187426 }

~0.5% gas increase

**Comment by danoctavian** [comment]
locking both requestUnstake and _processFirstUnstakeRequest during the migration because it can alter the structure of the list and disrupt the iteration (low probability though)

File: contracts/modules/staking/PooledStaking.sol [hunk 2]
```diff
@@ -138,8 +138,8 @@ contract PooledStaking is MasterAware, IPooledStaking {
   // contract address => ContractRewards
   mapping(address => ContractReward) public accumulatedRewards;
 
-  uint public REWARD_ROUND_DURATION;
-  uint public REWARD_ROUNDS_START;
+  uint public REWARD_ROUND_DURATION = 7 days;
+  uint public REWARD_ROUNDS_START = 1600074000;
```
Comment (shark0der) [inline contracts/modules/staking/PooledStaking.sol:142]: These are misleading. We're using a proxy so these will be unset in a proxy.

File: contracts/modules/staking/PooledStaking.sol [hunk 3]
```diff
@@ -995,61 +997,108 @@ contract PooledStaking is MasterAware, IPooledStaking {
...
+    uint next = firstId;
+
+    iterationsLeft = maxIterations;
+    while (!finished && iterationsLeft > 0) {
 
-      delete rewards[firstReward];
-      firstReward++;
+      iterationsLeft--;
+      UnstakeRequest storage unstakeRequest = unstakeRequests[next];
+      unstakeRequest.unstakeAt = unstakeRequest.unstakeAt - 60 days;
```
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1076]: should use SafeMath here

File: contracts/modules/staking/PooledStaking.sol [hunk 4]
```diff
@@ -994,62 +982,67 @@ contract PooledStaking is MasterAware, IPooledStaking {
...
 
-  bytes32 private constant MIGRATION_LAST_ID_POSITION = keccak256("nexusmutual.pooledstaking.MIGRATION_LAST_ID_POINTER");
+  function migratePendingUnstakesToNewLockTime(uint iterations) external {
 
-  function migrateRewardsToAccumulatedRewards(uint maxIterations) external returns (bool finished, uint iterationsLeft)  {
-    require(firstReward != 0, "Nothing to migrate");
+    uint migrationStatus;
+    bytes32 migrationStatusSlot = keccak256("nexusmutual.pooledstaking.LOCK_TIME_MIGRATION_STAGE");
+    assembly { migrationStatus := sload(migrationStatusSlot) }
+    require(migrationStatus == 0, 'PooledStaking: Migration finished');
```
Comment (roxdanila) [inline contracts/modules/staking/PooledStaking.sol:999]: nit: make it explicit what migration this is

File: test/fork/migrateTo30DaysLockTime.js [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    const [funder] = accounts;
+
+    for (const member of boardMembers) {
+      console.log(`Topping up ${member}`);
+      await web3.eth.sendTransaction({ from: funder, to: member, value: ether('100000') });
+      await unlock(member);
+    }
+
+    const REWARD_ROUND_DURATION = await ps.REWARD_ROUND_DURATION();
+    const REWARD_ROUNDS_START = await ps.REWARD_ROUNDS_START();
```
Comment (roxdanila) [inline test/fork/migrateTo30DaysLockTime.js:127]: don't need these

File: test/fork/migrateTo30DaysLockTime.js [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    const storedNewPSAddress = await psProxy.implementation();
+    assert.equal(storedNewPSAddress, newPSAddress);
+
+    const latestPS = await PooledStaking.at(await master.getLatestAddress(hex('PS')));
+
+    const NEW_REWARD_ROUND_DURATION = await latestPS.REWARD_ROUND_DURATION();
+    const NEW_REWARD_ROUNDS_START = await latestPS.REWARD_ROUNDS_START();
+
+    assert.equal(NEW_REWARD_ROUND_DURATION.toString(), REWARD_ROUND_DURATION.toString());
+    assert.equal(NEW_REWARD_ROUNDS_START.toString(), REWARD_ROUNDS_START.toString());
```
Comment (roxdanila) [inline test/fork/migrateTo30DaysLockTime.js:155]: same as above

File: test/fork/migrateTo30DaysLockTime.js [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+      const before = processedUnstakes[i];
+      const expectedUnstakeAt = before.unstakeAt.toNumber() - 60 * day;
+      assert(expectedUnstakeAt <= now);
+    }
+
+    for (let i = 0; i < unstakesBefore.length; i++) {
+
+      const before = unstakesBefore[i];
+      const after = unstakesAfter[i];
+      assert.equal(before.amount.toString(), before.amount.toString());
```
Comment (roxdanila) [inline test/fork/migrateTo30DaysLockTime.js:235]: uhm :)

**Review by roxdanila — APPROVED** [review]
minor comments, looks good overall

File: test/fork/migrateTo30DaysLockTime.js (hunk 7 above)
Comment (danoctavian) [reply test/fork/migrateTo30DaysLockTime.js:235]: oops

File: test/fork/migrateTo30DaysLockTime.js [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+const NXMaster = artifacts.require('NXMaster');
+const NXMToken = artifacts.require('NXMToken');
+const Governance = artifacts.require('Governance');
+const PooledStaking = artifacts.require('PooledStaking');
+const TokenFunctions = artifacts.require('TokenFunctions');
+const UpgradeabilityProxy = artifacts.require('UpgradeabilityProxy');
+
+const upgradeProxyImplementationCategoryId = 5;
+const newContractAddressUpgradeCategoryId = 29;
+const addNewInternalContractCategoryId = 34;
```
Comment (cleanunicorn) [inline test/fork/migrateTo30DaysLockTime.js:20]: These 2 variables don't seem to be used anywhere.

**Pull request merged**